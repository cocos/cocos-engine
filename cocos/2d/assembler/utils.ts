/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Color, Mat4, clamp } from '../../core';
import { RenderData } from '../renderer/render-data';
import { IBatcher } from '../renderer/i-batcher';
import { Node } from '../../scene-graph/node';
import { FormatInfos } from '../../gfx';

const m = new Mat4();

export function fillMeshVertices3D (node: Node, renderer: IBatcher, renderData: RenderData, color: Color) {
    const chunk = renderData.chunk;
    const dataList = renderData.data;
    const vData = chunk.vb;
    const vertexCount = renderData.vertexCount;

    node.getWorldMatrix(m);

    let vertexOffset = 0;
    for (let i = 0; i < vertexCount; i++) {
        const vert = dataList[i];
        const x = vert.x;
        const y = vert.y;
        let rhw = m.m03 * x + m.m07 * y + m.m15;
        rhw = rhw ? Math.abs(1 / rhw) : 1;
        vData[vertexOffset + 0] = (m.m00 * x + m.m04 * y + m.m12) * rhw;
        vData[vertexOffset + 1] = (m.m01 * x + m.m05 * y + m.m13) * rhw;
        vData[vertexOffset + 2] = (m.m02 * x + m.m06 * y + m.m14) * rhw;
        Color.toArray(vData, color, vertexOffset + 5);
        vertexOffset += 9;
    }

    // fill index data
    const bid = chunk.bufferId;
    const vid = chunk.vertexOffset;
    const meshBuffer = chunk.meshBuffer;
    const ib = chunk.meshBuffer.iData;
    let indexOffset = meshBuffer.indexOffset;
    for (let i = 0, count = vertexCount / 4; i < count; i++) {
        const start = vid + i * 4;
        ib[indexOffset++] = start;
        ib[indexOffset++] = start + 1;
        ib[indexOffset++] = start + 2;
        ib[indexOffset++] = start + 1;
        ib[indexOffset++] = start + 3;
        ib[indexOffset++] = start + 2;
    }
    meshBuffer.indexOffset += renderData.indexCount;
    meshBuffer.setDirty();
}

export function updateOpacity (renderData: RenderData, opacity: number) {
    const vfmt = renderData.vertexFormat;
    const vb = renderData.chunk.vb;
    let attr; let format; let stride;
    // Color component offset
    let offset = 0;
    for (let i = 0; i < vfmt.length; ++i) {
        attr = vfmt[i];
        format = FormatInfos[attr.format];
        if (format.hasAlpha) {
            stride = renderData.floatStride;
            if (format.size / format.count === 1) {
                const alpha = ~~clamp(Math.round(opacity * 255), 0, 255);
                // Uint color RGBA8
                for (let color = offset; color < vb.length; color += stride) {
                    vb[color] = ((vb[color] & 0xffffff00) | alpha) >>> 0;
                }
            } else if (format.size / format.count === 4) {
                // RGBA32 color, alpha at position 3
                for (let alpha = offset + 3; alpha < vb.length; alpha += stride) {
                    vb[alpha] = opacity;
                }
            }
        }
        offset += format.size >> 2;
    }
}
