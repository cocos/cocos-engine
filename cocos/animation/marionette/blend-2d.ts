/*
 Copyright (c) 2022-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { assertIsTrue, Vec2, Vec3 } from '../../core';

/**
 * Blends given samples using simple directional algorithm.
 * @param weights Result weights of each sample.
 * @param samples Every samples' parameter.
 * @param input Input parameter.
 */
export const blendSimpleDirectional = (() => {
    const CACHE_NORMALIZED_SAMPLE = new Vec2();

    const CACHE_BARYCENTRIC_SOLUTIONS: EquationResolutions = { wA: 0, wB: 0 };

    return function blendSimpleDirectional (weights: number[], samples: readonly Vec2[], input: Readonly<Vec2>) {
        assertIsTrue(weights.length === samples.length);

        if (samples.length === 0) {
            return;
        }

        if (samples.length === 1) {
            weights[0] = 1.0;
            return;
        }

        if (Vec2.strictEquals(input, Vec2.ZERO)) {
            const iCenter = samples.findIndex((sample) => Vec2.strictEquals(sample, Vec2.ZERO));
            if (iCenter >= 0) {
                weights[iCenter] = 1.0;
            } else {
                weights.fill(1.0 / samples.length);
            }
            return;
        }

        // Finds out the sector the input point locates
        let iSectorStart = -1;
        let iSectorEnd = -1;
        let iCenter = -1;
        let lhsCosAngle = Number.NEGATIVE_INFINITY;
        let rhsCosAngle = Number.NEGATIVE_INFINITY;
        const { x: inputX, y: inputY } = input;
        for (let iSample = 0; iSample < samples.length; ++iSample) {
            const sample = samples[iSample];
            if (Vec2.equals(sample, Vec2.ZERO)) {
                iCenter = iSample;
                continue;
            }

            const sampleNormalized = Vec2.normalize(CACHE_NORMALIZED_SAMPLE, sample);
            const cosAngle = Vec2.dot(sampleNormalized, input);
            const sign = sampleNormalized.x * inputY - sampleNormalized.y * inputX;
            if (sign > 0) {
                if (cosAngle >= rhsCosAngle) {
                    rhsCosAngle = cosAngle;
                    iSectorStart = iSample;
                }
            } else if (cosAngle >= lhsCosAngle) {
                lhsCosAngle = cosAngle;
                iSectorEnd = iSample;
            }
        }

        let centerWeight = 0.0;
        if (iSectorStart < 0 || iSectorEnd < 0) {
            // Input fall at vertex.
            centerWeight = 1.0;
        } else {
            const { wA, wB } = solveBarycentric(samples[iSectorStart], samples[iSectorEnd], input, CACHE_BARYCENTRIC_SOLUTIONS);
            let w1 = 0.0;
            let w2 = 0.0;
            const sum = wA + wB;
            if (sum > 1) {
                // Input fall at line C-A or C-B but not beyond C
                w1 = wA / sum;
                w2 = wB / sum;
            } else if (sum < 0) {
                // Input fall at line C-A or C-B but beyond A or B
                w1 = 0.0;
                w2 = 0.0;
                centerWeight = 1.0;
            } else {
                // Inside triangle
                w1 = wA;
                w2 = wB;
                centerWeight = 1.0 - sum;
            }
            weights[iSectorStart] = w1;
            weights[iSectorEnd] = w2;
        }

        // Center influence part
        if (centerWeight > 0.0) {
            if (iCenter >= 0) {
                weights[iCenter] = centerWeight;
            } else {
                const average = centerWeight / weights.length;
                for (let i = 0; i < weights.length; ++i) {
                    weights[i] += average;
                }
            }
        }
    };
})();

/**
 * Validates the samples if they satisfied the requirements of simple directional algorithm.
 * @param samples Samples to validate.
 * @returns Issues the samples containing.
 */
export function validateSimpleDirectionalSamples (samples: ReadonlyArray<Vec2>): SimpleDirectionalSampleIssue[] {
    const nSamples = samples.length;
    const issues: SimpleDirectionalSampleIssue[] = [];
    const sameDirectionValidationFlag = new Array<boolean>(samples.length).fill(false);
    samples.forEach((sample, iSample) => {
        if (sameDirectionValidationFlag[iSample]) {
            return;
        }
        let sameDirectionSamples: number[] | undefined;
        for (let iCheckSample = 0; iCheckSample < nSamples; ++iCheckSample) {
            const checkSample = samples[iCheckSample];
            if (Vec2.equals(sample, checkSample, 1e-5)) {
                (sameDirectionSamples ??= []).push(iCheckSample);
                sameDirectionValidationFlag[iCheckSample] = true;
            }
        }
        if (sameDirectionSamples) {
            sameDirectionSamples.unshift(iSample);
            issues.push(new SimpleDirectionalIssueSameDirection(sameDirectionSamples));
        }
    });
    return issues;
}

export type SimpleDirectionalSampleIssue = SimpleDirectionalIssueSameDirection;

/**
 * Simple directional issue representing some samples have same(or very similar) direction.
 */
export class SimpleDirectionalIssueSameDirection {
    public constructor (public samples: readonly number[]) { }
}

/**
 * Cartesian Gradient Band Interpolation.
 * @param weights
 * @param thresholds
 * @param value
 */
export function sampleFreeformCartesian (weights: number[], thresholds: readonly Vec2[], value: Readonly<Vec2>) {
    sampleFreeform(weights, thresholds, value, getGradientBandCartesianCoords);
}

/**
 * Polar Gradient Band Interpolation.
 * @param weights
 * @param thresholds
 * @param value
 */
export function sampleFreeformDirectional (weights: number[], thresholds: readonly Vec2[], value: Readonly<Vec2>) {
    sampleFreeform(weights, thresholds, value, getGradientBandPolarCoords);
}

function sampleFreeform (weights: number[], samples: readonly Vec2[], value: Readonly<Vec2>, getGradientBandCoords: GetGradientBandCoords) {
    weights.fill(0.0);
    const pIpInput = new Vec2(0, 0);
    const pIJ = new Vec2(0, 0);
    let sumInfluence = 0.0;
    const nSamples = samples.length;
    for (let iSample = 0; iSample < nSamples; ++iSample) {
        let influence = Number.MAX_VALUE;
        let outsideHull = false;
        for (let jSample = 0; jSample < nSamples; ++jSample) {
            if (iSample === jSample) {
                continue;
            }
            getGradientBandCoords(samples[iSample], samples[jSample], value, pIpInput, pIJ);
            const t = 1 - Vec2.dot(pIpInput, pIJ) / Vec2.lengthSqr(pIJ);
            if (t < 0) {
                outsideHull = true;
                break;
            }
            influence = Math.min(influence, t);
        }
        if (!outsideHull) {
            weights[iSample] = influence;
            sumInfluence += influence;
        }
    }
    if (sumInfluence > 0) {
        weights.forEach((influence, index) => weights[index] = influence / sumInfluence);
    }
}

interface EquationResolutions {
    wA: number;
    wB: number;
}

/**
 * Solves the barycentric coordinates of `p` within triangle (0, `a`, `b`).
 * @param a Triangle vertex.
 * @param b Triangle vertex.
 * @param p Input vector.
 * @param resolutions The barycentric coordinates of `a` and `b`.
 * @returns
 */
function solveBarycentric (
    a: Readonly<Vec2>,
    b: Readonly<Vec2>,
    p: Readonly<Vec2>,
    resolutions: EquationResolutions,
) {
    // Let P = p - 0, A = a - 0, B = b - 0,
    // wA = (P x B) / (A x B)
    // wB = (P x A) / (B x A)
    const det = Vec2.cross(a, b);
    if (!det) {
        resolutions.wA = 0.0;
        resolutions.wB = 0.0;
    } else {
        resolutions.wA = Vec2.cross(p, b) / det;
        resolutions.wB = Vec2.cross(p, a) / -det;
    }
    return resolutions;
}

type GetGradientBandCoords = (point: Readonly<Vec2>, pI: Readonly<Vec2>, pJ: Readonly<Vec2>, pIpInput: Vec2, pIpJ: Vec2) => void;

const getGradientBandCartesianCoords: GetGradientBandCoords = (pI, pJ, input, pIpInput, pIpJ) => {
    Vec2.subtract(pIpInput, input, pI);
    Vec2.subtract(pIpJ, pJ, pI);
};

const getGradientBandPolarCoords = ((): GetGradientBandCoords => {
    const axis = new Vec3(0, 0, 0); // buffer for axis
    const tmpV3 = new Vec3(0, 0, 0); // buffer for temp vec3
    const pQueriedProjected = new Vec3(0, 0, 0); // buffer for pQueriedProjected
    const pi3 = new Vec3(0, 0, 0); // buffer for pi3
    const pj3 = new Vec3(0, 0, 0); // buffer for pj3
    const pQueried3 = new Vec3(0, 0, 0); // buffer for pQueried3
    return (pI, pJ, input, pIpInput, pIpJ) => {
        let aIJ = 0.0;
        let aIQ = 0.0;
        let angleMultiplier = 2.0;
        Vec3.set(pQueriedProjected, input.x, input.y, 0.0);
        if (Vec2.equals(pI, Vec2.ZERO)) {
            aIJ = Vec2.angle(input, pJ);
            aIQ = 0.0;
            angleMultiplier = 1.0;
        } else if (Vec2.equals(pJ, Vec2.ZERO)) {
            aIJ = Vec2.angle(input, pI);
            aIQ = aIJ;
            angleMultiplier = 1.0;
        } else {
            aIJ = Vec2.angle(pI, pJ);
            if (aIJ <= 0.0) {
                aIQ = 0.0;
            } else if (Vec2.equals(input, Vec2.ZERO)) {
                aIQ = aIJ;
            } else {
                Vec3.set(pi3, pI.x, pI.y, 0);
                Vec3.set(pj3, pJ.x, pJ.y, 0);
                Vec3.set(pQueried3, input.x, input.y, 0);
                Vec3.cross(axis, pi3, pj3);
                Vec3.projectOnPlane(pQueriedProjected, pQueried3, axis);
                aIQ = Vec3.angle(pi3, pQueriedProjected);
                if (aIJ < Math.PI * 0.99) {
                    if (Vec3.dot(Vec3.cross(tmpV3, pi3, pQueriedProjected), axis) < 0) {
                        aIQ = -aIQ;
                    }
                }
            }
        }
        const lenPI = Vec2.len(pI);
        const lenPJ = Vec2.len(pJ);
        const deno = (lenPJ + lenPI) / 2;
        Vec2.set(pIpJ, (lenPJ - lenPI) / deno, aIJ * angleMultiplier);
        Vec2.set(pIpInput, (Vec3.len(pQueriedProjected) - lenPI) / deno, aIQ * angleMultiplier);
    };
})();
