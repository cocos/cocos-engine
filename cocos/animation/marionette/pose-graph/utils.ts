/**
 * Tells if the weight is too small so that it can be treated as 0.
 * @param weight The weight.
 * @returns True if it can be treated as 0.
 */
export function isIgnorableWeight (weight: number): boolean {
    return weight < 1e-5;
}
