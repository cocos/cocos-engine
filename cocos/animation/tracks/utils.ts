import type { Curve } from './track';

export function maskIfEmpty<T extends Curve> (curve: T) {
    return curve.keyFramesCount === 0 ? undefined : curve;
}

export interface Range {
    min: number;
    max: number;
}
