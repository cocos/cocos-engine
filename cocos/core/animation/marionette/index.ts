export { InvalidTransitionError, VariableNotDefinedError } from './errors';
export { AnimationGraph, LayerBlending, isAnimationTransition, StateMachine, SubStateMachine } from './animation-graph';
export type { Transition, AnimationTransition, Layer, State, Variable } from './animation-graph';
export { BinaryCondition, UnaryCondition, TriggerCondition } from './condition';
export type { Condition } from './condition';
export type { Value } from './variable';
export { MotionState } from './motion-state';
export { ClipMotion } from './clip-motion';
export type { AnimationBlend } from './animation-blend';
export { AnimationBlendDirect, AnimationBlendDirectItem } from './animation-blend-direct';
export { AnimationBlend1D, AnimationBlend1DItem } from './animation-blend-1d';
export { AnimationBlend2D, AnimationBlend2DItem } from './animation-blend-2d';
export { AnimationController } from './animation-controller';
export type { ClipStatus, TransitionStatus, StateStatus } from './animation-controller';
export { VariableType } from './parametric';
export type { BindableNumber, BindableBoolean } from './parametric';
export { SkeletonMask } from '../skeleton-mask';
export { StateMachineComponent } from './state-machine-component';

export { __getDemoGraphs } from './__tmp__/get-demo-graphs';
