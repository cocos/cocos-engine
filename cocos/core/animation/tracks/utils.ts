import type { Curve } from './track';

export function maskIfEmpty<T extends Curve> (curve: T) {
    return curve.empty ? undefined : curve;
}

export interface Range {
    min: number;
    max: number;
}
