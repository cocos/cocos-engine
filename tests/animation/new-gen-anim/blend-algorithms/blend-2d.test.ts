import { Vec2 } from "../../../../cocos/core";
import { sampleFreeformDirectional, blendSimpleDirectional } from "../../../../cocos/animation/marionette/blend-2d";
import '../../../utils/matcher-deep-close-to';

const EXPECT_NUM_DIGITS = 5;

describe('Simple directional 2D', () => {
    function calcSimpleDirectional (samples: readonly Vec2[], input: Vec2) {
        const weights = new Array(samples.length).fill(0);
        blendSimpleDirectional(weights, samples, input);
        return weights;
    }

    test('Zero or one sample', () => {
        expect(() => calcSimpleDirectional([], new Vec2(3.14, 6.18))).not.toThrow();
        
        expect(calcSimpleDirectional([new Vec2(-12.3, 45.6)], new Vec2(78.9, 3.14))).toBeDeepCloseTo([1.0], EXPECT_NUM_DIGITS);
    });
    
    test('Input direction is zero', () => {
        // If there is also zero sample, that sample owns all weights.
        expect(calcSimpleDirectional([
            new Vec2(-0.3, 0.4),
            new Vec2(+1.0, 0.2),
            new Vec2(0.0, 0.0),
            new Vec2(0.0, -1.0),
            new Vec2(0.0, +1.0),
        ], Vec2.ZERO)).toBeDeepCloseTo([
            0.0,
            0.0,
            1.0,
            0.0,
            0.0,
        ], EXPECT_NUM_DIGITS);

        // Otherwise, weights are averaged for all samples.
        expect(calcSimpleDirectional([
            new Vec2(-0.3, 0.4),
            new Vec2(+1.0, 0.2),
            new Vec2(0.0, -1.0),
            new Vec2(0.0, +1.0),
        ], Vec2.ZERO)).toBeDeepCloseTo([
            0.25,
            0.25,
            0.25,
            0.25,
        ], EXPECT_NUM_DIGITS);
    });

    test('Input within a sector', () => {
        // Points can be visualized at: https://www.desmos.com/calculator/rapuusa3g1

        const calcEdgeLinePoint = (from: Vec2, to: Vec2, ratio: number) => {
            return Vec2.scaleAndAdd(
                new Vec2(),
                from,
                Vec2.subtract(new Vec2(), to, from),
                ratio,
            );
        };

        const sampleTriangleVertexA = new Vec2(0.2, 0.7);
        const sampleTriangleVertexB = new Vec2(0.3, 0.4);
        const sampleOutOfSector1 = new Vec2(0.0576,0.583);
        const sampleOutOfSector2 = new Vec2(0.243,-0.22);
        const inputInsideTriangle = new Vec2(0.15, 0.4);
        const inputOnEdgeABRatio = 0.34;
        const inputOnEdgeAB = calcEdgeLinePoint(sampleTriangleVertexA, sampleTriangleVertexB, inputOnEdgeABRatio);
        const inputOnEdgeCARatio = 0.711;
        const inputOnEdgeCA = calcEdgeLinePoint(Vec2.ZERO, sampleTriangleVertexA, inputOnEdgeCARatio);
        const inputOnEdgeCAExtended = calcEdgeLinePoint(Vec2.ZERO, sampleTriangleVertexA, 1.3);
        const inputOnEdgeCBExtended = calcEdgeLinePoint(Vec2.ZERO, sampleTriangleVertexB, -0.4);
        const inputInsideSectorOutSideTriangle = new Vec2(0.2765,0.682);

        const samples = [
            sampleTriangleVertexA,
            sampleTriangleVertexB,
            sampleOutOfSector1,
            sampleOutOfSector2,
        ];

        const samplesWithCentralSample = [
            ...samples,
            Vec2.ZERO,
        ];

        const weightsOfSample = ({
            sampleTriangleVertexA = 0.0,
            sampleTriangleVertexB = 0.0,
            sampleOutOfSector1 = 0.0,
            sampleOutOfSector2 = 0.0,
        }: Partial<Record<
            | 'sampleTriangleVertexA'
            | 'sampleTriangleVertexB'
            | 'sampleOutOfSector1'
            | 'sampleOutOfSector2'
        , number>>) => {
            return [
                sampleTriangleVertexA,
                sampleTriangleVertexB,
                sampleOutOfSector1,
                sampleOutOfSector2,
            ];
        };

        const weightsOfSampleWithCentralSample = ({ sampleCentral = 0.0, ...other }: Parameters<typeof weightsOfSample>[0] & {
            sampleCentral?: number;
        }) => {
            return [
                ...weightsOfSample(other),
                sampleCentral,
            ];
        };

        // Outside any sector.
        expect(calcSimpleDirectional([
            sampleTriangleVertexA,
            sampleTriangleVertexB,
        ], new Vec2(0.4, 0.4))).toBeDeepCloseTo([
            0.5,
            0.5,
        ], EXPECT_NUM_DIGITS);

        // Inside triangle, without central sample
        // central weight are averaged to all
        expect(calcSimpleDirectional(samples, inputInsideTriangle)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: 0.46154 + 0.34615 / samples.length,
            sampleTriangleVertexB: 0.19231 + 0.34615 / samples.length,
            sampleOutOfSector1: 0.34615 / samples.length,
            sampleOutOfSector2: 0.34615 / samples.length,
        }), EXPECT_NUM_DIGITS);

        // Inside triangle, with central sample
        expect(calcSimpleDirectional(samplesWithCentralSample, inputInsideTriangle)).toBeDeepCloseTo(weightsOfSampleWithCentralSample({
            sampleTriangleVertexA: 0.46154,
            sampleTriangleVertexB: 0.19231,
            sampleCentral: 0.34615,
        }), EXPECT_NUM_DIGITS);

        // Fall at vertex
        expect(calcSimpleDirectional(samples, sampleTriangleVertexA)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: 1.0,
        }), EXPECT_NUM_DIGITS);

        // Fall at edge(between center vertex and vertex A), without central sample
        expect(calcSimpleDirectional(samples, inputOnEdgeCA)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: inputOnEdgeCARatio + (1.0 - inputOnEdgeCARatio) / samples.length,
            sampleTriangleVertexB: (1.0 - inputOnEdgeCARatio) / samples.length,
            sampleOutOfSector1: (1.0 - inputOnEdgeCARatio) / samples.length,
            sampleOutOfSector2: (1.0 - inputOnEdgeCARatio) / samples.length,
        }), EXPECT_NUM_DIGITS);

        // Fall at edge(between center vertex and vertex A), with central sample
        expect(calcSimpleDirectional(samplesWithCentralSample, inputOnEdgeCA)).toBeDeepCloseTo(weightsOfSampleWithCentralSample({
            sampleCentral: 1.0 - inputOnEdgeCARatio,
            sampleTriangleVertexA: inputOnEdgeCARatio,
        }), EXPECT_NUM_DIGITS);

        // Fall on edge(between vertex A, B)
        expect(calcSimpleDirectional(samples, inputOnEdgeAB)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: (1.0 - inputOnEdgeABRatio),
            sampleTriangleVertexB: inputOnEdgeABRatio,
        }), EXPECT_NUM_DIGITS);

        // Fall on line(C -> A) but not edge(C -> A), beyond A.
        // CentBarycentric coordinates are normalized.
        expect(calcSimpleDirectional(samples, inputOnEdgeCAExtended)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: 1.0,
        }), EXPECT_NUM_DIGITS);

        // Fall on line(C -> B) but not edge(C -> B), not beyond C, without central sample.
        // Only central point got weight, but there is no central sample. So averaged to every sample.
        expect(calcSimpleDirectional(samples, inputOnEdgeCBExtended)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: 1.0 / samples.length,
            sampleTriangleVertexB: 1.0 / samples.length,
            sampleOutOfSector1: 1.0 / samples.length,
            sampleOutOfSector2: 1.0 / samples.length,
        }), EXPECT_NUM_DIGITS);

        // Fall on line(C -> B) but not edge(C -> B), not beyond C, with central sample.
        // Only central point got weight, 
        expect(calcSimpleDirectional(samplesWithCentralSample, inputOnEdgeCBExtended)).toBeDeepCloseTo(weightsOfSampleWithCentralSample({
            sampleCentral: 1.0,
        }), EXPECT_NUM_DIGITS);

        // Inside sector, but outside triangle.
        // Similar with "fall on line" case: centBarycentric coordinates are normalized.
        expect(calcSimpleDirectional(samples, inputInsideSectorOutSideTriangle)).toBeDeepCloseTo(weightsOfSample({
            sampleTriangleVertexA: 0.6219,
            sampleTriangleVertexB: 0.3781,
        }), EXPECT_NUM_DIGITS);
    });
});