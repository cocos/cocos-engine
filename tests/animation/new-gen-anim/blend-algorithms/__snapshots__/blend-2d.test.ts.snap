// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polar space gradient brand interpolation Algorithm The input's direction is between two different direction examples The examples have different magnitude 1`] = `
Object {
  "Equal to the bigger magnitude": Object {
    "ccw": 0.3168596289437383,
    "cw": 0.6502105566924706,
    "origin": 0.03292981436379107,
  },
  "Equal to the middle magnitude": Object {
    "ccw": 0.2989265859181596,
    "cw": 0.5902714914805254,
    "origin": 0.110801922601315,
  },
  "Equal to the minor magnitude": Object {
    "ccw": 0.28366638688732265,
    "cw": 0.5392660896894322,
    "origin": 0.17706752342324497,
  },
  "Greater than the bigger magnitude": Object {
    "ccw": 0.30491192209599804,
    "cw": 0.695088077904002,
  },
  "Less than both magnitudes": Object {
    "ccw": 0.16486765633781594,
    "cw": 0.15973168002542656,
    "origin": 0.6754006636367574,
  },
}
`;

exports[`Polar space gradient brand interpolation Algorithm The input's direction is between two different direction examples The examples have same magnitude 1`] = `
Object {
  "Greater than the mag": Object {
    "ccw": 0.33333333333333337,
    "cw": 0.6666666666666667,
  },
  "Greater than the mag again": Object {
    "ccw": 0.33333333333333337,
    "cw": 0.6666666666666667,
  },
  "Input at just the mag": Object {
    "ccw": 0.3205352705404605,
    "cw": 0.641070541080921,
    "origin": 0.03839418837861866,
  },
  "Less then the mag.": Object {
    "ccw": 0.1648595779760476,
    "cw": 0.1965831709644441,
    "origin": 0.6385572510595082,
  },
}
`;

exports[`Polar space gradient brand interpolation Algorithm There are multiple examples along the input's direction 1`] = `
Object {
  "Input at 1/3 minor mag": Object {
    "ccw": 0.08602309330508458,
    "cw": 0.05841787524451817,
    "origin": 0.5703726924902929,
    "same_dir_example_1": 0.2851863389601044,
  },
  "Input at bigger mag example": Object {
    "same_dir_example_2": 0.9999999767855615,
  },
  "Input at minor mag example": Object {
    "same_dir_example_1": 0.9999999717541427,
  },
  "Input between the examples": Object {
    "same_dir_example_1": 0.6666666589285206,
    "same_dir_example_2": 0.3333333410714793,
  },
  "Input beyond bigger mag example": Object {
    "same_dir_example_2": 1,
  },
}
`;

exports[`Polar space gradient brand interpolation Algorithm There is and is only one example along the input's direction The adjacent examples have different mag 1`] = `
Object {
  "Input at 1/3 mag": Object {
    "cw": 0.02054480721915246,
    "origin": 0.6529701236555586,
    "the_dir": 0.32648506912528885,
  },
  "Input at the example": Object {
    "the_dir": 0.9999999848019046,
  },
  "Input beyond the example": Object {
    "the_dir": 1,
  },
}
`;

exports[`Polar space gradient brand interpolation Algorithm There is and is only one example along the input's direction The adjacent examples have same mag 1`] = `
Object {
  "Input at 1/3 mag": Object {
    "origin": 0.6666666616996132,
    "the_dir": 0.33333333830038686,
  },
  "Input at the example": Object {
    "the_dir": 0.9999999850988393,
  },
  "Input beyond the example": Object {
    "the_dir": 1,
  },
}
`;

exports[`Polar space gradient brand interpolation Algorithm There is and is only one example along the input's direction the adjacent examples have different mag, the direction is on axis 1`] = `
Object {
  "Input at 1/3 mag": Object {
    "origin": 0.6666666616996132,
    "the_dir": 0.33333333830038686,
  },
  "Input at the example": Object {
    "the_dir": 0.9999999850988393,
  },
  "Input beyond the example": Object {
    "cw_adjacent": 0.31394171978396535,
    "the_dir": 0.6860582802160347,
  },
}
`;

exports[`Polar space gradient brand interpolation There is more than 180° between two adjacent examples points: Between 1, 2 1`] = `
Object {
  "Equal magnitude": Object {
    "example_1": 0.7939990314904765,
    "example_2": 0.19849978329369805,
    "origin": 0.007501185215825347,
  },
  "Longer magnitude": Object {
    "example_1": 0.7999999795094334,
    "example_2": 0.20000002049056667,
  },
  "Shorter magnitude": Object {
    "example_1": 0.6616280659241143,
    "example_2": 0.1666666856394138,
    "origin": 0.171705248436472,
  },
}
`;

exports[`Polar space gradient brand interpolation There is more than 180° between two adjacent examples points: Between 2, 3 1`] = `
Object {
  "Equal magnitude": Object {
    "example_2": 0.7443741424354346,
    "example_3": 0.24812467177667524,
    "origin": 0.007501185787890177,
  },
  "Longer magnitude": Object {
    "example_2": 0.7500000320165141,
    "example_3": 0.24999996798348595,
  },
  "Shorter magnitude": Object {
    "example_2": 0.6218666935060354,
    "example_3": 0.20728886243978442,
    "origin": 0.17084444405418012,
  },
}
`;

exports[`Polar space gradient brand interpolation There is more than 180° between two adjacent examples points: In the more-than-180 range, in the middle 1`] = `
Object {
  "Equal magnitude": Object {
    "example_1": 0.09981031755884037,
    "example_2": 0.8982930455035218,
    "origin": 0.0018966369376378289,
  },
  "Longer magnitude": Object {
    "example_1": 0.09999998121697959,
    "example_2": 0.9000000187830204,
  },
  "Shorter magnitude": Object {
    "example_1": 0.0909090739747374,
    "example_2": 0.7258907473599425,
    "origin": 0.18320017866531998,
  },
}
`;

exports[`Polar space gradient brand interpolation There is more than 180° between two adjacent examples points: In the more-than-180 range, tending to 1 1`] = `
Object {
  "Equal magnitude": Object {
    "example_3": 0.8400637213479694,
    "origin": 0.15993627865203064,
  },
  "Longer magnitude": Object {
    "example_3": 1,
  },
  "Shorter magnitude": Object {
    "example_3": 0.6720509770783756,
    "origin": 0.3279490229216244,
  },
}
`;

exports[`Polar space gradient brand interpolation There is more than 180° between two adjacent examples points: In the more-than-180 range, tending to 3 1`] = `
Object {
  "Equal magnitude": Object {
    "example_3": 0.8400637213479694,
    "origin": 0.15993627865203064,
  },
  "Longer magnitude": Object {
    "example_3": 1,
  },
  "Shorter magnitude": Object {
    "example_3": 0.6720509770783756,
    "origin": 0.3279490229216244,
  },
}
`;
